spec.bs

```diff
@@ -447,6 +448,83 @@ To <dfn>build an interest group permissions url</dfn> given a [=origin=] |ownerO
+    [=interest group/owner=].
+
+<div algorithm='maintenance'>
+To <dfn>perform storage maintenance</dfn>:
+
+1. Let |ownersAndExpiry| be a new [=ordered map=] whose [=map/keys=] are [=origins=]
+  [=interest group/owner=] and [=map/values=] are [=moments=] [=interest group/expiry=].
+1. [=list/For each=] |ig| of the user agent's [=interest group set=]:
+  1. Let |owner| be |ig|'s [=interest group/owner=].
+  1. If |ownersAndExpiry|[|owner|] [=map/exists=], then [=map/set=] |ownersAndExpiry|[|owner|] to
+    |ig|'s [=interest group/expiry=] if it comes after |ownersAndExpiry|[|owner|].
+  1. Otherwise, [=map/set=] |ownersAndExpiry|[|owner|] to |ig|'s [=interest group/expiry=].
+1. If [=map/size=] of |ownersAndExpiry| |i| &gt; 2000, then set |ownersAndExpiry| to
+  |ownersAndExpiry| [=map/sorted in descending order=] with |a| being less than |b| if |a|'s
+  [=map/value=] comes before |b|'s [=map/value=].
+
+  Note: Sorting owners based on their expiry in descending order so that if interest groups of some
+    owners need to be removed, the interest groups of owners expiring soonest are the ones to be
+    removed first.
+
+1. Let |owners| be the [=map/get the keys|keys=] of |ownersAndExpiry|.
+1. [=list/For each=] |i| in [=the range=] from 0 to the [=set/size=] of |owners|, exclusive:
+  1. If |i| &ge; 2000, then [=list/remove=] [=interest groups=] from the user agent's
+    [=interest group set=] whose [=interest group/owner=] is |owners|[|i|], and [=iteration/continue=].
+  1. Let |regularIgs| be a [=list=] of [=regular interest groups=] in the user agent's
+    [=interest group set=] whose [=interest group/owner=] is |owners|[|i|].
+  1. If [=list/size=] of |regularIgs| &gt; 1000, then [=clear excess interest groups=] with
+    |regularIgs|, |owners|[|i|], and 1000.
+  1. Let |negativeIgs| be a [=list=] of [=negative interest groups=] in the user agent's
+    [=interest group set=] whose [=interest group/owner=] is |owners|[|i|].
+  1. If [=list/size=] of |regularIgs| &gt; 20000, then [=Clear excess interest groups=] with
+    |negativeIgs|, |owners|[|i|], and 20000.
+1. TODO: remove expired IGs.
+1. [=list/For each=] |owner| of |owners|:
+  1. Let |igs| be a [=list=] of [=interest groups=] in the user agent's [=interest group set=] whose
+    [=interest group/owner=] is |owner|, [=list/sorted in descending order=] with |a| being less
+    than |b| if |a|[=interest group/expiry=] comes before |b|[=interest group/expiry=].
+  1. Let |cumulativeSize| be 0.
+  1. [=list/For each=] |ig| of |igs|:
+    1. If the sum of |cumulativeSize| and |ig|'s [=interest group/estimated size=]
+      &gt; <code>10\*1024\*1024</code>, then [=list/remove=] |ig| from the user
+      agent's [=interest group set=].
+    1. Otherwise, increment |cumulativeSize| by |ig|'s [=interest group/estimated size=].
+
+</div>
+
+<div algorithm>
+To <dfn>clear excess interest groups</dfn> with a [=list=] of [=interest groups=] |igs|,

```

Comment:
@domfarolino This algorithm basically implements what the [`Note`](https://github.com/WICG/turtledove/pull/859/files#diff-6f5a1d8263b0b0c42e2716ba5750e3652e359532647ac934c1c70086ae3ceddaR518) says below. Do you think I need to actually spec this sorting and then deleting algorithm like what it currently does, or just some sentences like the what the Note says is OK?

Reply:
Yeah I think including the algorithms is a good move. I'd make the `Note` a more complete sentence though. It kind of doesn't make as much sense starting with `Sorting [...]`